File: 백준/Silver/2805. 나무 자르기/나무 자르기.js
let fs = require("fs");

let line = fs.readFileSync("/dev/stdin").toString().trim().split("\n");

let [N, M] = line[0].split(" ").map(Number);
let trees = line[1].split(" ").map(Number);

let low = 0;
let high = trees.reduce((a, b) => Math.max(a, b));


let answer = 0;
while (low <= high) {
  let mid = parseInt((low + high) / 2);

  let sum = 0;
  trees.forEach(el => {
    if (el > mid) {
      sum += (el - mid);
    }
  })

  if (sum < M) {
    high = mid - 1;
  } else {
    answer = mid;
    low = mid + 1;
  }
}

console.log(answer)

File: 백준/Silver/2805. 나무 자르기/나무 자르기.java
import java.io.*;
import java.util.*;


public class Main {
    static int N, M;
    static int[] arr;
    static int answer = 0;

    static void binarySearch(int s, int e) {

        int mid;
        long sum;
        while (s <= e) {
            mid = (s + e) / 2;

            sum = 0;
            for (int i = 0; i < N; i++) {
                if(arr[i] >= mid) sum += arr[i] - mid;
            }

            if (sum < M) {
                e = mid - 1;
            } else {
                s = mid + 1;
                answer = Math.max(answer, mid);
            }
        }

    }

    public static void main(String[] args) throws  IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        arr = new int[N];
        st = new StringTokenizer(br.readLine());
        int max = 0;
        for (int i = 0; i < N; i++) {
            arr[i] = Integer.parseInt(st.nextToken());
            max = Math.max(max, arr[i]);
        }

        binarySearch(0, max);


        bw.write(String.valueOf(answer));
        bw.flush();
        bw.close();
        br.close();
    }
}